### PR TITLE
[StructMech] Automatically create adjoint settings for linear static cases

### PR DESCRIPTION
**Pull request by armingeiser**
For the linear static case, the process of setting up an adjoint response function has been simplified.

The response function (on python level) acts as a top level entity that knows about both, primal and adjoint analysis.
This is why it is possible to automatically create the parameters for the adjoint analysis by just slightly modifying the primal parameters file.
Also the HDF5 process is not necessary in this case, because the primal results can be directly transferred on python level.

The more flexible way of manually defining the adjoint parameters is still possible though.

### CODE REVIEW

File: applications/StructuralMechanicsApplication/python_scripts/structural_response.py [hunk 1]
```diff
@@ -363,4 +374,69 @@ def _SynchronizeAdjointFromPrimal(self):
...
+
+            # Output process:
+            # TODO how to add the output process? How find out about the variables?
+            if adjoint_parameters.Has("output_configuration"):
+                Logger.PrintInfo("> Output process is removed for adjoint analysis. To enable it define adjoint_parameters yourself.")
+                adjoint_parameters.RemoveValue("output_configuration")
+
+            # sensitivity settings
+            adjoint_parameters["solver_settings"].AddValue("sensitivity_settings", self.response_settings["sensitivity_settings"])
+            self.response_settings.RemoveValue("sensitivity_settings")
```
Comment (dbaumgaertner) [inline applications/StructuralMechanicsApplication/python_scripts/structural_response.py]: Why do you remove the sensitivity settings here?

File: applications/StructuralMechanicsApplication/tests/response_function_tests/adjoint_displacement_response_parameters.json [hunk 2]
```diff
@@ -16,6 +16,13 @@
         "traced_node_id"       : 5,
         "traced_dof"        : "DISPLACEMENT_Z",
         "primal_settings"  : "rectangular_plate_parameters.json",
-        "adjoint_settings"  : "rectangular_plate_adjoint_parameters.json"
+        "adjoint_settings"  : "auto",
```
Comment (dbaumgaertner) [inline applications/StructuralMechanicsApplication/tests/response_function_tests/adjoint_displacement_response_parameters.json:19]: Looks very clean now!

File: applications/StructuralMechanicsApplication/tests/structural_response_function_test_factory.py [hunk 3]
```diff
@@ -150,12 +143,11 @@ class TestStrainEnergyResponseFunction(StructuralResponseFunctionTestFactory):
     def test_execution(self):
         self.current_model = KratosMultiphysics.Model()
         self._calculate_response_and_gradient()
-        self.assertAlmostEqual(self.value, 0.00606275111915477)
+        self.assertAlmostEqual(self.value, 0.6062751119154768)
```
Comment (dbaumgaertner) [inline applications/StructuralMechanicsApplication/tests/structural_response_function_test_factory.py:143]: Why did you change this and the below values by a factor 100? Where did this come from?

**Review by dbaumgaertner — COMMENTED** [review]
Looks good to me. I very much welcome the independence from hdf5 for the standard static case and like the simplification in the settings.

File: applications/StructuralMechanicsApplication/tests/structural_response_function_test_factory.py (hunk 3 above)
Comment (armingeiser) [reply applications/StructuralMechanicsApplication/tests/structural_response_function_test_factory.py:143]: Previously there where two primal parameter files, one for the old version of the strain energy and one for the adjoint workflow with hdf5. 
For some reason they had different load magnitudes. Now as none of the tests is using hdf5 anymore i could use the same primal parameters everywhere, thats why i had to adapt the test results for the old strain energy response.

File: applications/StructuralMechanicsApplication/python_scripts/structural_response.py (hunk 1 above)
Comment (armingeiser) [reply applications/StructuralMechanicsApplication/python_scripts/structural_response.py]: because technically they are not part of the settings that are passed down to the cpp response.
The sensitivity_settings are only needed in the sensitivity builder a.k.a adjoint postprocess.

File: applications/StructuralMechanicsApplication/python_scripts/structural_response.py [hunk 4]
```diff
@@ -307,6 +309,15 @@ def InitializeSolutionStep(self):
...
         # this needs to be improved, also the response value should be calculated on the PRIMAL modelpart!!
         self.adjoint_analysis.time = self.adjoint_analysis._GetSolver().AdvanceInTime(self.adjoint_analysis.time)
+
+        # Put primal solution on adjoint model - for "auto" setting, else it has to be done by the user e.g. using hdf5 process
+        if self.response_settings["adjoint_settings"].GetString() == "auto":
+            Logger.PrintInfo("> Transfer primal state to adjoint model part.")
+            for primal_node, adjoint_node in zip(self.primal_model_part.Nodes, self.adjoint_model_part.Nodes):
+                for variable in self.primal_state_variables:
+                    values_to_transfer = primal_node.GetSolutionStepValue(variable)
+                    adjoint_node.SetSolutionStepValue(variable, values_to_transfer)
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/structural_response.py]: Please use the function from variable-utils for this

File: applications/StructuralMechanicsApplication/python_scripts/structural_response.py [hunk 5]
```diff
@@ -363,4 +374,69 @@ def _SynchronizeAdjointFromPrimal(self):
...
+
+            if not solver_settings.Has("move_mesh_flag"):
+                solver_settings.AddEmptyValue("move_mesh_flag")
+            solver_settings["move_mesh_flag"].SetBool(False)
+
+            if not solver_settings.Has("scheme_settings"):
+                tmp = solver_settings.AddEmptyValue("scheme_settings")
+                if not tmp.Has("scheme_type"):
+                    tmp.AddEmptyValue("scheme_type")
+            solver_settings["scheme_settings"]["scheme_type"].SetString("adjoint_structural")
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/structural_response.py:422]: those things are quite manual

I think here you could use the assigndefaults-function once that is merged

**Review by philbucher — COMMENTED** [review]
minor

File: applications/StructuralMechanicsApplication/python_scripts/structural_response.py (hunk 5 above)
Comment (armingeiser) [reply applications/StructuralMechanicsApplication/python_scripts/structural_response.py:422]: No, here i really have to overwrite the setting in case it already exists.

File: applications/StructuralMechanicsApplication/python_scripts/structural_response.py (hunk 4 above)
Comment (armingeiser) [reply applications/StructuralMechanicsApplication/python_scripts/structural_response.py]: don't know why i always forget this. thanks!

File: applications/StructuralMechanicsApplication/python_scripts/structural_response.py (hunk 1 above)
Comment (dbaumgaertner) [reply applications/StructuralMechanicsApplication/python_scripts/structural_response.py]: Still, it seems like not a very good practice here. One would expect that a single settings object always just adds settings but never removes, otherwise, why should it be there in the first place, if it is just needed locally. But I do not have the complete picture here, so there might be indeed good reasons to do so.

Comment (armingeiser) [reply applications/StructuralMechanicsApplication/python_scripts/structural_response.py]: Maybe we discuss it in person.

**Comment by armingeiser** [comment]
ping

**Comment by armingeiser** [comment]
I resolved the conflicts, @MFusseder or @dbaumgaertner can you please approve or request further changes.

**Review by MFusseder — DISMISSED** [review]

**Review by MFusseder — APPROVED** [review]
